File: Question_11_20/answers_cpp/answer_18.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
#include <math.h>

int main(int argc, const char* argv[]){
  cv::Mat img = cv::imread("imori.jpg", cv::IMREAD_COLOR);

  int width = img.rows;
  int height = img.cols;

  // gray scale
  cv::Mat gray = cv::Mat::zeros(height, width, CV_8UC1);
  
  for (int j=0; j<height; j++){
    for (int i=0; i<width; i++){
      gray.at<uchar>(j, i) = (int)((float)img.at<cv::Vec3b>(j,i)[0] * 0.0722 + \
				   (float)img.at<cv::Vec3b>(j,i)[1] * 0.7152 + \
				   (float)img.at<cv::Vec3b>(j,i)[2] * 0.2126);
    }
  }

  cv::Mat out = cv::Mat::zeros(height, width, CV_8UC1);

  // kernel
  int k_size = 3;
  int p = floor(k_size / 2);

  double k[k_size][k_size] = {{-2, -1, 0}, {-1, 1, 1}, {0, 1, 2}};
  
  // filtering
  double v = 0;
  
  for (int j = 0; j < height; j++){
    for (int i = 0; i < width; i++){
      v = 0;
      for (int _j = -p; _j < p+1; _j++){
	for (int _i = -p; _i < p+1; _i++){
	  if (((j+_j) >= 0) && ((i+_i) >= 0) && ((j+_j) < height) && ((i+_i) < width)){
	    v += gray.at<uchar>(j+_j, i+_i) * k[_j+p][_i+p];
	  }
	}
      }
      if(v > 255){
	v = 255;
      }
      if(v < 0){
	v = 0;
      }
      
      out.at<uchar>(j, i) = v;
    }
  }
  
  //cv::imwrite("out.jpg", out);
  cv::imshow("answer", out);
  cv::waitKey(0);
  cv::destroyAllWindows();

  return 0;
}

File: Question_11_20/answers/answer_18.py
import cv2
import numpy as np

# Gray scale
def BGR2GRAY(img):
	b = img[:, :, 0].copy()
	g = img[:, :, 1].copy()
	r = img[:, :, 2].copy()

	# Gray scale
	out = 0.2126 * r + 0.7152 * g + 0.0722 * b
	out = out.astype(np.uint8)

	return out

# emboss filter
def emboss_filter(img, K_size=3):
	H, W, C = img.shape

	# zero padding
	pad = K_size // 2
	out = np.zeros((H + pad * 2, W + pad * 2), dtype=np.float)
	out[pad: pad + H, pad: pad + W] = gray.copy().astype(np.float)
	tmp = out.copy()

	# emboss kernel
	K = [[-2., -1., 0.],[-1., 1., 1.], [0., 1., 2.]]

	# filtering
	for y in range(H):
		for x in range(W):
			out[pad + y, pad + x] = np.sum(K * (tmp[y: y + K_size, x: x + K_size]))

	out = np.clip(out, 0, 255)
	out = out[pad: pad + H, pad: pad + W].astype(np.uint8)

	return out


# Read image
img = cv2.imread("imori.jpg").astype(np.float)

# grayscale
gray = BGR2GRAY(img)

# emboss filtering
out = emboss_filter(gray, K_size=3)


# Save result
cv2.imwrite("out.jpg", out)
cv2.imshow("result", out)
cv2.waitKey(0)
cv2.destroyAllWindows()
